org.eclipse.ui.genericeditor/src/org/eclipse/ui/internal/genericeditor/folding/IndentFoldingStrategy.java

```diff
@@ -243,10 +248,15 @@ public void reconcile(DirtyRegion dirtyRegion, IRegion subRegion) {
 				Integer lastLineForKeyword = null;
 				int line = endLine;
 				for (line = endLine; line >= 0; line--) {
-					int lineOffset = document.getLineOffset(line);
-					String delim = document.getLineDelimiter(line);
-					int lineLength = document.getLineLength(line) - (delim != null ? delim.length() : 0);
-					String lineContent = document.get(lineOffset, lineLength);
+					thisDocument = document;
+					if (thisDocument == null) {
+						// Exit as soon as possible if uninstalled
+						return;
```

Comment:
Why repeating the block here? Can't we just assume that the earlier setting of thisDocument line 226 is enough?

Reply:
good catch

Reply:
@mickaelistria This repeated block is added into a cycle to make it stop calculations as soon as possible as we don't need the results anymore.

It looks like `document` becomes `null` after a call to `uninstall` method from a different thread:
```
	public void uninstall() {
		setDocument(null);

		if (viewer != null) {
			viewer.removeProjectionListener(this);
			viewer = null;
		}

		projectionDisabled();
	}
```
So we're getting caught between  `setDocument(null)` and `projectionDisabled` calls - the check for `projectionAnnotationModel != null` in  the beginning of `reconcile` method tells me that document is already `null` while `projectionAnnotationModel` is still not. 
So we can have the same at any time during the `reconcile` method work despite the fact that `uninstall` method in fact consist of only 3 (three) method calls. And as such, the appearance of `null` document inside the `for`-loop  is quite possible after it was not `null` in the first check.  Also, if during the cycle we get `uninstall` invoked then we don't need to continue the processing cycle because we don't need its results anymore. IMHO, it's better to stop and exit it as soon as possible.

Reply:
> This repeated block is added into a cycle to make it stop calculations as soon as possible as we don't need the results anymore.

OK, but you don't need to repeat the assignment then.
You could even include the condition in the `for` loop about `for (line = endLine; line >= 0 && this.document != null; line--) {`

Reply:
Yes, you're right.